backend/src/datasources/userDatasource.js

```diff
@@ -0,0 +1,94 @@
+import { DataSource } from 'apollo-datasource'
+import { UserInputError } from 'apollo-server'
+import bcrypt from 'bcrypt'
+import { generateAccessToken } from '../utils/tokenGenerator'
+import { User } from '../models/user'
+
+export class UserDataSource extends DataSource {
+  constructor (users = []) {
+    super()
+    this.users = users
+  }
+
+  initialize ({ context }) {
+    this.context = context
+  }
+
+  async getUserById (id) {
+    if (id) {
+      return this.users.find((user) => user.id === id)
+    } else {
+      throw new UserInputError(
+        'No matching user found. Please check your input'
+      )
+    }
+  }
+
+  async getUserByName (name) {
+    if (name) {
+      return this.users.find((user) => user.name === name)
+    } else {
+      throw new UserInputError(
+        'No matching user found. Please check your input'
+      )
+    }
+  }
+
+  async getUserByEmail (email) {
+    if (email) {
+      return this.users.find((user) => user.email === email)
+    } else {
+      throw new UserInputError(
+        'No matching user found. Please check your input'
+      )
+    }
+  }
+
+  async getAll () {
+    return this.users
+  }
+
+  async removeUser (id) {
+    this.users = this.users.filter((user) => {
```

Comment:
Consider the following change:
```suggestion
    this.users = this.users.filter((user) => user.id !== id);
```